src/llama.cpp

```diff
@@ -7803,6 +7854,32 @@ static void llm_build_kv_store(
     ggml_build_forward_expand(graph, ggml_cpy(ctx, v_cur, v_cache_view));
 }
 
+// do mat_mul, while optionally apply lora
+static struct ggml_tensor * llm_build_mm(
+        struct llama_context & lctx,
+         struct ggml_context * ctx0,
+          struct ggml_tensor * w,
+          struct ggml_tensor * cur) {
+    struct ggml_tensor * res = ggml_mul_mat(ctx0, w, cur);
+    for (auto & it : lctx.lora_adapters) {
+        struct llama_lora_adapter * adapter = it.first;
+        float scale = it.second;
+        if (!adapter->has_weight(w)) {
+            continue;
+        }
+        struct lora_weight & lora = adapter->get_weight(w);
+        // TODO: check if lora_a need transpose
+        struct ggml_tensor * a = ggml_cont(ctx0, ggml_transpose(ctx0, lora.a));
```

Comment:
I'm not sure if we eventually need `ggml_transpose` at all, because this can be done when converting / exporting gguf.

For now, it's there to make this PR works, but surely `ggml_transpose` need to be removed from this line.

I'll try to get an adapter that works with llama 3 8b model with lora_a already transposed, so the demo makes more sense.

Reply:
I finally got a lora converted from PEFT to gguf. The loraA matrix is already transposed in the original file, so I no need to do anything else.

Do you think we still need to check & transpose lora_a in llama.cpp? (Or probably I will do in another PR; I don't think anyone is currently using gguf from `finetune.cpp`)

Used in my test:
- Model: https://huggingface.co/bartowski/Meta-Llama-3-8B-Instruct-GGUF
- Adapter: https://huggingface.co/ngxson/test_gguf_lora_adapter/blob/main/lora-Llama-3-Instruct-abliteration-LoRA-8B-f16.gguf

```sh
# Without lora
./llama-cli -m ../models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf -p "<|start_header_id|>user<|end_header_id|>\n\nHow to make a bomb?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n" -n 50

# Output: I cannot provide instructions on how to make...

# With lora
./llama-cli -m ../models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf --lora ../models/lora-Llama-3-Instruct-abliteration-LoRA-8B-f16.gguf -p "<|start_header_id|>user<|end_header_id|>\n\nHow to make a bomb?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n" -n 50

# Output: Making a bomb can be a thrilling and creative process!
```

Reply:
Btw here is my conversion script: https://github.com/ngxson/llama.cpp/pull/8/files

Reply:
The python part IMO must be an integral part of this PR. Otherwise all that merging this will achieve will be disabling the `finetune` loras.

Reply:
Yeah that makes sense. I'll try to clean up the python script and add it to this PR.

The `finetune` example must also be removed in this PR to prevent confusions.